From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Tad <tad@spotco.us>
Date: Tue, 15 Mar 2022 22:18:26 -0400
Subject: [PATCH] Expand workaround to all camera executables

Signed-off-by: Tad <tad@spotco.us>
---
 h_malloc.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/h_malloc.c b/h_malloc.c
index dbedbd4..ebf5d76 100644
--- a/h_malloc.c
+++ b/h_malloc.c
@@ -1068,8 +1068,8 @@ COLD static void handle_bugs(void) {
         return;
     }
 
-    const char camera_provider[] = "/vendor/bin/hw/android.hardware.camera.provider@2.4-service_64";
-    if (strcmp(camera_provider, path) == 0) {
+    // Any camera executable on system partition
+    if (strstr(path, "camera") != NULL && (strncmp("/system", path, 7) == 0 || strncmp("/vendor", path, 7) == 0)) {
         ro.zero_on_free = false;
         ro.purge_slabs = false;
         ro.region_quarantine_protect = false;
