From c58e3bb0f5ee76b82c4249cd7b4432b7fb711add Mon Sep 17 00:00:00 2001
From: Daniel Micay <danielmicay@gmail.com>
Date: Tue, 6 Dec 2016 17:11:48 -0500
Subject: [PATCH] use build tools 25.0.1

---
 app/build.gradle | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/app/build.gradle b/app/build.gradle
index 7ae8ca4..a81160d 100644
--- a/app/build.gradle
+++ b/app/build.gradle
@@ -146,7 +146,7 @@ if (!hasProperty('sourceDeps')) {
 
 android {
     compileSdkVersion 24
-    buildToolsVersion '25.0.0'
+    buildToolsVersion '25.0.1'
     useLibrary 'org.apache.http.legacy'
 
     buildTypes {
