From 74a364df2dfe05460cffd081ac3bab672faf346c Mon Sep 17 00:00:00 2001
From: Tectas <adrian.stoeckl@gmail.com>
Date: Thu, 20 Nov 2014 21:28:16 +0100
Subject: [PATCH] arch/arm/boot/dts: Add more frequencies for UC

---
 arch/arm/boot/dts/msm8974.dtsi | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/arch/arm/boot/dts/msm8974.dtsi b/arch/arm/boot/dts/msm8974.dtsi
index fc19c18e716..83ead25e266 100644
--- a/arch/arm/boot/dts/msm8974.dtsi
+++ b/arch/arm/boot/dts/msm8974.dtsi
@@ -1569,6 +1569,7 @@
 		compatible = "qcom,msm-cpufreq";
 		qcom,cpufreq-table =
 			<  300000  300000  572 >,
+			<  345600  300000  572 >,
 			<  422400  422400 1144 >,
 			<  652800  499200 1525 >,
 			<  729600  576000 2342 >,
@@ -1579,9 +1580,16 @@
 			< 1267200 1267200 4684 >,
 			< 1497600 1497600 4684 >,
 			< 1574400 1574400 6103 >,
+			< 1651200 1574400 6103 >,
 			< 1728000 1651200 6103 >,
+			< 1804800 1651200 6103 >,
 			< 1958400 1728000 7102 >,
+			< 2035200 1728000 7102 >,
+			< 2112000 1728000 7102 >,
+			< 2188800 1728000 7102 >,
 			< 2265600 1728000 7102 >,
+			< 2342400 1728000 7102 >,
+			< 2419200 1728000 7102 >,
 			< 2457600 1728000 7102 >,
 			< 2534400 1728000 7102 >,
 			< 2611200 1728000 7102 >,
