From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Daniel Micay <danielmicay@gmail.com>
Date: Tue, 19 Oct 2021 15:17:55 +0300
Subject: [PATCH] stop auto-granting location to system browsers

---
 .../behavior/BrowserRoleBehavior.java         | 29 -------------------
 1 file changed, 29 deletions(-)

diff --git a/PermissionController/role-controller/java/com/android/role/controller/behavior/BrowserRoleBehavior.java b/PermissionController/role-controller/java/com/android/role/controller/behavior/BrowserRoleBehavior.java
index 0261e1eee..685a5b004 100644
--- a/PermissionController/role-controller/java/com/android/role/controller/behavior/BrowserRoleBehavior.java
+++ b/PermissionController/role-controller/java/com/android/role/controller/behavior/BrowserRoleBehavior.java
@@ -54,11 +54,6 @@ public class BrowserRoleBehavior implements RoleBehavior {
             .addCategory(Intent.CATEGORY_BROWSABLE)
             .setData(Uri.fromParts("http", "", null));
 
-    private static final List<String> SYSTEM_BROWSER_PERMISSIONS = Arrays.asList(
-            android.Manifest.permission.ACCESS_COARSE_LOCATION,
-            android.Manifest.permission.ACCESS_FINE_LOCATION
-    );
-
     @Nullable
     @Override
     public String getFallbackHolderAsUser(@NonNull Role role, @NonNull UserHandle user,
@@ -132,30 +127,6 @@ public class BrowserRoleBehavior implements RoleBehavior {
         return new ArrayList<>(packageNames);
     }
 
-    @Override
-    public void grantAsUser(@NonNull Role role, @NonNull String packageName,
-            @NonNull UserHandle user, @NonNull Context context) {
-        // @see com.android.server.pm.permission.DefaultPermissionGrantPolicy
-        //      #grantDefaultPermissionsToDefaultBrowser(java.lang.String, int)
-        if (SdkLevel.isAtLeastS()) {
-            if (PackageUtils.isSystemPackageAsUser(packageName, user, context)) {
-                Permissions.grantAsUser(packageName, SYSTEM_BROWSER_PERMISSIONS, false, false,
-                        true, false, false, user, context);
-            }
-        }
-    }
-
-    @Override
-    public void revokeAsUser(@NonNull Role role, @NonNull String packageName,
-            @NonNull UserHandle user, @NonNull Context context) {
-        if (SdkLevel.isAtLeastT()) {
-            if (PackageUtils.isSystemPackageAsUser(packageName, user, context)) {
-                Permissions.revokeAsUser(packageName, SYSTEM_BROWSER_PERMISSIONS, true, false,
-                        false, user, context);
-            }
-        }
-    }
-
     @Override
     public boolean isVisibleAsUser(@NonNull Role role, @NonNull UserHandle user,
             @NonNull Context context) {
