From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dmitry Muhomor <muhomor.dmitry@gmail.com>
Date: Thu, 5 Oct 2023 15:16:24 +0300
Subject: [PATCH] mark hardened_malloc as bp2build_available

---
 android/allowlists/allowlists.go | 1 +
 1 file changed, 1 insertion(+)

diff --git a/android/allowlists/allowlists.go b/android/allowlists/allowlists.go
index b9d26f884..34c83e00c 100644
--- a/android/allowlists/allowlists.go
+++ b/android/allowlists/allowlists.go
@@ -160,6 +160,7 @@ var (
 		"external/guava":                           Bp2BuildDefaultTrueRecursively,
 		"external/gwp_asan":                        Bp2BuildDefaultTrueRecursively,
 		"external/hamcrest":                        Bp2BuildDefaultTrueRecursively,
+		"external/hardened_malloc":                 Bp2BuildDefaultTrueRecursively,
 		"external/icu":                             Bp2BuildDefaultTrueRecursively,
 		"external/icu/android_icu4j":               Bp2BuildDefaultFalse, // java rules incomplete
 		"external/icu/icu4j":                       Bp2BuildDefaultFalse, // java rules incomplete
