From bb788ab88084c09a34444f3e12d4f6cd1cc7398b Mon Sep 17 00:00:00 2001
From: Tad <tad@spotco.us>
Date: Sun, 18 Dec 2016 17:38:28 -0500
Subject: [PATCH] Update build tools

---
 app/build.gradle | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/app/build.gradle b/app/build.gradle
index cbba2e6..88b97c7 100644
--- a/app/build.gradle
+++ b/app/build.gradle
@@ -36,7 +36,7 @@ dependencies {
 
 android {
     compileSdkVersion 24
-    buildToolsVersion '24.0.1'
+    buildToolsVersion '25.0.1'
 
     defaultConfig {
         archivesBaseName = "FDroidPrivilegedExtension"
-- 
2.9.3

