From 63c497725740bfb6c673456d38f14d8c6030c991 Mon Sep 17 00:00:00 2001
From: Daniel Micay <danielmicay@gmail.com>
Date: Tue, 13 Sep 2016 22:05:56 -0400
Subject: [PATCH] use -fwrapv when signed overflow checking is off

---
 core/config_sanitizers.mk | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/core/config_sanitizers.mk b/core/config_sanitizers.mk
index d3adee5ae2..83988bb603 100644
--- a/core/config_sanitizers.mk
+++ b/core/config_sanitizers.mk
@@ -435,3 +435,9 @@ ifneq ($(findstring fsanitize,$(my_cflags)),)
     endif
   endif
 endif
+
+ifeq ($(filter signed-integer-overflow integer undefined,$(my_sanitize)),)
+  ifeq ($(filter -ftrapv,$(my_cflags)),)
+    my_cflags += -fwrapv
+  endif
+endif
