From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: anupritaisno1 <www.anuprita804@gmail.com>
Date: Sun, 13 Sep 2020 09:35:41 +0000
Subject: [PATCH] make hardened malloc available to apexes

---
 apex/apex.go | 1 +
 1 file changed, 1 insertion(+)

diff --git a/apex/apex.go b/apex/apex.go
index f12c517e9..d45a6eaca 100644
--- a/apex/apex.go
+++ b/apex/apex.go
@@ -2907,6 +2907,7 @@ func makeApexAvailableBaseline() map[string][]string {
 		"libdebuggerd_handler_fallback",
 		"libdl_static",
 		"libjemalloc5",
+		"libhardened_malloc",
 		"liblinker_main",
 		"liblinker_malloc",
 		"liblz4",
