From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Daniel Micay <danielmicay@gmail.com>
Date: Sat, 7 Sep 2019 19:27:52 -0400
Subject: [PATCH] disable symbol ordering for now

---
 libc/Android.bp | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/libc/Android.bp b/libc/Android.bp
index e7ca82f92..6c753091b 100644
--- a/libc/Android.bp
+++ b/libc/Android.bp
@@ -1643,7 +1643,7 @@ cc_library {
         versions: ["10000"],
     },
 
-    symbol_ordering_file: "symbol_ordering",
+    //symbol_ordering_file: "symbol_ordering",
 }
 
 genrule {
